### PR TITLE
GODRIVER-2119 implement methods required to use  as a map key

### PR DESCRIPTION
**Pull request by Luberry**
[GODRIVER-2119](https://jira.mongodb.org/browse/GODRIVER-2119)

Implements the [TextMarshaler](https://cs.opensource.google/go/go/+/go1.16.7:src/encoding/encoding.go;l=36) and [TextUnmarshaler](https://cs.opensource.google/go/go/+/go1.16.7:src/encoding/encoding.go;l=46) methods on `primitive.ObjectID` to allow it to be used as a map key.

### CODE REVIEW

**Review by benjirewis — COMMENTED** [review]
Thanks again for your contribution @Luberry ! Just some small comments.

**Review by benjirewis — APPROVED** [review]
LGTM! I'll wait for the rest of the team to review; thanks.

File: bson/primitive/objectid.go
Comment (kevinAlbs) [inline bson/primitive/objectid.go:106]: `ObjectID.UnmarshalJSON()` accepts multiple input formats:

1. the string "null"
2. a 12 byte raw representation
3. an empty string (which unmarshals to a NilObjectID)
4. a 24 character string hex representation
5. a representation of the [canonical extended JSON format](https://github.com/mongodb/specifications/blob/df6be82f865e9b72444488fd62ae1eb5fca18569/source/extended-json.rst#conversion-table): `{"$oid": <ObjectId bytes as 24-character, big-endian hex string>}`

@Luberry would a preferable alternative be to use `ObjectIDFromHex`? It only accepts a 24 character hex string. That may better align with the intent of the `UnmarshalText` interface.

**Review by kevinAlbs — COMMENTED** [review]
Thank you for the contribution @Luberry! I have a question about an alternative for `UnmarshalText`.

File: bson/primitive/objectid.go
Comment (Luberry) [reply bson/primitive/objectid.go:106]: I thought of that, I figured this would make sure we we handle extracting them the same as is done with json. So that if someone had a file expecting any other possible formats they could handle it. I'm not too attached either way. And ultimately it's up to you all.

File: bson/primitive/objectid.go
Comment (matthewdale) [inline bson/primitive/objectid.go:97]: Consider adding a variable assignment that asserts that `primitive.ObjectID` implements `encoding.TextMarshaler` and `encoding.TextUnmarshaler`.

E.g.
```go
var _ encoding.TextMarshaler = ObjectID{}
var _ encoding.TextUnmarshaler = ObjectID{}
```

File: bson/primitive/objectid_test.go
Comment (matthewdale) [inline bson/primitive/objectid_test.go:191]: Consider incorporating this test function as a test case in `TestObjectID_UnmarshalJSON`.

File: bson/primitive/objectid.go
Comment (matthewdale) [reply bson/primitive/objectid.go:106]: I agree with @kevinAlbs's suggestion to limit the implementation to strictly unmarshal from the 24-char string hex representation (i.e. use `ObjectIDFromHex()`). I think that avoids having to support unexpected formats without reducing the usefulness of the feature.

**Review by matthewdale — APPROVED** [review]
Approved, with a few suggestions, and pending resolution of the `UnmarshalText` conversation.

File: bson/primitive/objectid.go
Comment (Luberry) [reply bson/primitive/objectid.go:106]: alright, ill implement that.

File: bson/primitive/objectid_test.go
Comment (Luberry) [reply bson/primitive/objectid_test.go:191]: I did not add it in `TestObjectID_UnmarshalJSON` because i did not see a clean way of slotting it into the table tests.

**Review by kevinAlbs — APPROVED** [review]
LGTM, thank you again for the contribution @Luberry!

**Pull request merged**
